### PR TITLE
HDDS-7632. Add ability in Recon to track various Container Sizes across Ozone cluster.

### PR DESCRIPTION
**Pull request by ArafatKhan2198**
## What changes were proposed in this pull request?

The ContainerSizeCountTask class is a task that runs in the background to periodically scan the list of containers managed by the Storage Container Manager (SCM) and keep track of container sizes, binned into ranges. It stores this information in the Recon database in a table called `container_count_by_size`

This task runs on a specified interval and gets the list of containers from the SCM through the ContainerManager object. It then iterates through the list of containers and updates the count of containers in the appropriate size range in the container_count_by_size table. If the task is run for the first time, it will `reprocess` all the containers and update the counts for all size ranges. But once it runs again it will simply executes the `process` method that just applies delta updates to the `container_count_by_size` table. 

This task can be useful for getting an overview of the distribution of container sizes in the cluster. This information can be useful for capacity planning and resource allocation.
## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-7632?filter=-1

Document on how the ContainerSizeCountTask works and its code flow 
https://docs.google.com/document/d/1MoS6fEIHEjHsJKYs0cuDwvZLH9hnmVsevQq8xvswzfQ/edit?usp=sharing

Document on how we dynamically calculate the container sizes and alot a bin number for each size :-  https://docs.google.com/document/d/1itPS0dIzkF-tI2mGrWWCYDEdLQ4h8jRFsGMOxp_Meys/edit#

## How was this patch tested?
- Added basic unit tests for `Process()` and `Reprocess()` methods
- Tested it by manually by exposing the data through the `UtilizationEndpoint` and by populating the cluster with Data :- http://localhost:9888/api/v1/utilization/containerCount

image

image

**_After Balancing_** 
image

image

image

### CODE REVIEW

**Comment by kerneltime** [comment]
cc @dombizita @devmadhuu please take a look.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ReconTaskConfig.java [hunk 1]
```diff
@@ -66,4 +66,22 @@ public void setMissingContainerTaskInterval(Duration interval) {
...
+  )
+  private long containerSizeCountTaskInterval =
+      Duration.ofMinutes(1).toMillis();
+
+  public Duration getContainerSizeCountTaskInterval() {
+    return Duration.ofMillis(missingContainerTaskInterval);
+  }
+
+  public void setContainerSizeCountTaskInterval(Duration interval) {
+    this.missingContainerTaskInterval = interval.toMillis();
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ReconTaskConfig.java:84]: Why we are getting and setting missingContainerTaskInterval ? This is different task.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ReconTaskConfig.java [hunk 2]
```diff
@@ -66,4 +66,22 @@ public void setMissingContainerTaskInterval(Duration interval) {
...
 
+  @Config(key = "containercounttask.interval",
+      type = ConfigType.TIME,
+      defaultValue = "300s",
+      tags = { ConfigTag.RECON, ConfigTag.OZONE },
+      description = "The time interval of the periodic check for " +
+          "container size distribution in the cluster as reported "
+  )
+  private long containerSizeCountTaskInterval =
+      Duration.ofMinutes(1).toMillis();
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ReconTaskConfig.java:77]: Why setting 1 minute here when we are defining the config value default as 300s ?

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+  protected synchronized void run() {
+    try {
+      while (canRun()) {
+        wait(interval);
+
+        final List<ContainerInfo> containers = containerManager.getContainers();
+        Map<ContainerSizeCountKey, Long> containerSizeCountMap =
+            new HashMap<>();
+        for (int i = 0; i < containers.size(); i++) {
+          handlePutKeyEvent(containers.get(i), containerSizeCountMap);
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java:85]: Try to do profile and time test this with at least 100K containers, as we are deleting whole table below and doing fresh operation every 5 mins.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+  @Override
+  public String getTaskName() {
+    return "ContainerSizeCountTask";
+  }
+
+  /**
+   * Calculate and update the count of containers being tracked by
+   * containerSizeCountMap.
+   * Used by reprocess() and process().
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java:132]: Not sure if this comment is true. Don't see any reprocess or process method.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    this.dslContext = dslContext;
+    interval = reconTaskConfig.getContainerSizeCountTaskInterval().toMillis();
+  }
+
+
+  @Override
+  protected synchronized void run() {
+    try {
+      while (canRun()) {
+        wait(interval);
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java:79]: Try to refactor the code below this line in separate method, so that it can be triggered on any other event like when a new container being added on SCM side instead of waiting till "interval" time. Can be doable in new JIRA for above scenario, however refactor to new method below this line.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+          handlePutKeyEvent(containers.get(i), containerSizeCountMap);
+        }
+        // Truncate table before inserting new rows
+        int execute = dslContext.delete(CONTAINER_COUNT_BY_SIZE).execute();
+        LOG.info("Deleted {} records from {}", execute,
+            CONTAINER_COUNT_BY_SIZE);
+
+        writeCountsToDB(true, containerSizeCountMap);
+
+        LOG.info("Completed a 'reprocess' run of ContainerSizeCountTask.");
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java:94]: Correct this comment.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  /**
+   * Populate DB with the counts of container sizes calculated
+   * using the dao.
+   */
+  private void writeCountsToDB(boolean isDbTruncated,
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java:108]: Do we need this "isDBTruncated" flag if it is always true ?

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+  }
+
+  /**
+   * Calculate and update the count of containers being tracked by
+   * containerSizeCountMap.
+   * Used by reprocess() and process().
+   *
+   * @param containerInfo OmKey being updated for count
+   */
+  private void handlePutKeyEvent(ContainerInfo containerInfo,
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java:136]: Try to give better name, as there is no event we are capturing with SCM DB sync, here we are taking data from Recon container cache.

**Review by devmadhuu — CHANGES_REQUESTED** [review]

**Comment by ArafatKhan2198** [comment]
@devmadhuu @dombizita @smengcl Can you do a review for this please

**Review by dombizita — COMMENTED** [review]
thanks for working on this @ArafatKhan2198, I would like to know a bit more about the design choices of this task. as this is extending the `ReconScmTask`, the `run()` method needs to be overridden. as in that method we have this line: `final List<ContainerInfo> containers = containerManager.getContainers();`, so in each periodic run we will get all the containers and its info, right? Because of this I don't get the idea of `process` and `reprocess` (that is used in the `ReconOmTask`, because there in each run we will get only the delta of the `OMDBUpdateEvent`s, or all of them if there is some failure in the processing). So in this `ContainerSizeCountTask` I think we can have only a `processContainers()` method, which will handle the changes accordingly. let me know if I misunderstood something :)

**Comment by ArafatKhan2198** [comment]
@dombizita @kerneltime @smengcl @devmadhuu @xBis7 
Can you please do a review on this !!

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+
+import static org.hadoop.ozone.recon.schema.tables.ContainerCountBySizeTable.CONTAINER_COUNT_BY_SIZE;
+
+
+/**
+ * Class that scans the list of containers and keeps track of container sizes
+ * binned into ranges (1KB, 2Kb…,4MB,…1GB,…5GB…) to the Recon
+ * containerSize DB.
+ */
+public class ContainerSizeCountTask extends ReconScmTask {
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java:50]: This task should also be called on ICR from DN, so that most upto date ContainerSize Distribution information is with Recon. Pls check how it is being done for ContainerHealthTask and PipeLineSyncTask.

Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java:50]: Thats a good idea! I'll make the required changes !!

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/UtilizationEndpoint.java [hunk 10]
```diff
@@ -49,21 +55,24 @@ public class UtilizationEndpoint {
...
+    this.containerCountBySizeDao = containerCountBySizeDao;
   }
 
   /**
    * Return the file counts from Recon DB.
    * @return {@link Response}
    */
   @GET
-  @Path("/fileCount")
+  @Path("/pop")
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/UtilizationEndpoint.java:75]: If this change is impacting taking care on UI, whether we have created separate task to handle UI changes ? Pls mention and make sure that PR for this change and PR for UI should be merged together.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconDeadNodeHandler.java [hunk 11]
```diff
@@ -79,6 +85,7 @@ public void onMessage(final DatanodeDetails datanodeDetails,
       }
       containerHealthTask.triggerContainerHealthCheck();
       pipelineSyncTask.triggerPipelineSyncTask();
+      containerSizeCountTask.process(containerManager.getContainers());
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconDeadNodeHandler.java:88]: Have we verified if a node becomes dead, how container size will be impacted. Container may get closed on that DN, but some other replica for that container may be present in other DN in cluster. Pls check if we really need to call our task for DeadNode handler.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconStaleNodeHandler.java [hunk 12]
```diff
@@ -50,6 +55,7 @@ public void onMessage(final DatanodeDetails datanodeDetails,
     super.onMessage(datanodeDetails, publisher);
     try {
       pipelineSyncTask.triggerPipelineSyncTask();
+      containerSizeCountTask.process(containerManager.getContainers());
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconStaleNodeHandler.java:58]: Have we verified if a node becomes stale, how container size will be impacted. Container may get closed on that DN, but some other replica for that container may be present in other DN in cluster. Pls check if we really need to call our task for StaleNode handler.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconIncrementalContainerReportHandler.java [hunk 13]
```diff
@@ -99,5 +103,6 @@ public void onMessage(final IncrementalContainerReportFromDatanode report,
       }
     }
     containerManager.notifyContainerReportProcessing(false, success);
+    containerSizeCountTask.process(containerManager.getContainers());
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconIncrementalContainerReportHandler.java:106]: As we discussed, ICR handler get called for every DN, so we need to first check if our task map already contains the container, if yes, then it means , our task will not be impacted just for state as our task maintains count based on size. If container not present in our. map, then only we should call our task. Pls check how complex and feasible this to implement. When ICR handler gets called for deletion of containers, we should verify first after whether containerManager is having the container or not, if not then our task can be called to handle for deletion of container.

Please first write down all use case scenarios related to ICR why and when it is needed. Same for Stale and Dead Node handler.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/UtilizationEndpoint.java (hunk 10 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/UtilizationEndpoint.java:75]: This was by mistake! thanks for pointing it out !!

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconDeadNodeHandler.java (hunk 11 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconDeadNodeHandler.java:88]: For DeadDatanodeHandler we will be calling **destroyPipelines, closeContainers and removeContainerReplicas methods.** Hence since the container replicas are removed, the container size counts will also be affected. Therefore, we need to trigger the ContainerSizeCountTask.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconStaleNodeHandler.java (hunk 12 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconStaleNodeHandler.java:58]: The main function of the stale node handler is to ensure that the data stored in the pipelines associated with the stale node is not accessible anymore and also log the information about the stale node.

**Hence, since neither containers or their replicas are removed, the container size counts will not be affected. Therefore, we would not have to trigger the ContainerSizeCountTask.**

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconIncrementalContainerReportHandler.java (hunk 13 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconIncrementalContainerReportHandler.java:106]: Since If the container does not exist, it creates a new container and adds it to the ContainerManager hence we will have to trigger the ContainerSizeCountTask.

Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconIncrementalContainerReportHandler.java:106]: I have written in detail about how each of the three handlers work, in this google doc :- 

https://docs.google.com/document/d/1Wz-zeSoSPfdaVyw43GwAyvFNgt0djo5w44j8OjrxJ1I/edit#

**Comment by DaveTeng0** [comment]
Hey @ArafatKhan2198 ~ there were some CI tests failures. Please help take a look when you have some time, thanks!

**Comment by jojochuang** [comment]
Error:  Failures: 
Error:    TestContainerSizeCountTask.testProcess:99 expected: <2> but was: <1>

this test failure is most likely related to this PR. @ArafatKhan2198 please double check

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+   * in that size range within a Map.
+   *
+   * Used by decrementContainerSizeCount() and incrementContainerSizeCount()
+   *
+   * @param containerSize to calculate the upperSizeBound
+   */
+  private ContainerSizeCountKey getContainerSizeCountKey(
+      long containerSize) {
+    return new ContainerSizeCountKey(
+        // Using the FileSize UpperBound Calculator for now, we can replace it
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java:334]: Pls change the description, not to use FileSize upper bound

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/UtilizationEndpoint.java [hunk 15]
```diff
@@ -98,4 +107,37 @@ public Response getFileCounts(
...
+  /**
+   * Return the container size counts from Recon DB.
+   *
+   * @return {@link Response}
+   */
+  @GET
+  @Path("/containerCount")
+  public Response getContainerCounts(
+      @QueryParam(RECON_QUERY_CONTAINER_SIZE)
+          long containerSize
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/UtilizationEndpoint.java:120]: Pls change the name of parameter "containerSize" to "upperBound"

File: hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/scm/TestReconIncrementalContainerReportHandler.java [hunk 16]
```diff
@@ -133,6 +137,8 @@ public void testProcessICRStateMismatch()
           containerWithPipeline.getPipeline().getFirstNode();
       NodeManager nodeManagerMock = mock(NodeManager.class);
       when(nodeManagerMock.getNodeByUuid(any())).thenReturn(datanodeDetails);
+      ContainerSizeCountTask containerSizeCountTask =
+          mock(ContainerSizeCountTask.class);
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/scm/TestReconIncrementalContainerReportHandler.java:141]: Pls see if you can how easy to add assertion for change in size count

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/UtilizationEndpoint.java [hunk 17]
```diff
@@ -98,4 +111,44 @@ public Response getFileCounts(
...
+        Record1<Long> recordToFind =
+            dslContext.newRecord(
+                    CONTAINER_COUNT_BY_SIZE.CONTAINER_SIZE)
+                .value1(containerSizeUpperBound);
+        ContainerCountBySize record =
+            containerCountBySizeDao.findById(recordToFind.value1());
+        resultSet = record != null ?
+            Collections.singletonList(record) : Collections.emptyList();
+      } else {
+        // fetch all records having values greater than zero
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/UtilizationEndpoint.java:141]: why don't we fetch all the records, even if a container size has 0 occurance?

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconDeadNodeHandler.java [hunk 18]
```diff
@@ -44,17 +45,22 @@ public class ReconDeadNodeHandler extends DeadNodeHandler {
...
                               PipelineManager pipelineManager,
                               ContainerManager containerManager,
                               StorageContainerServiceProvider scmClient,
                               ContainerHealthTask containerHealthTask,
-                              PipelineSyncTask pipelineSyncTask) {
+                              PipelineSyncTask pipelineSyncTask,
+                              ContainerSizeCountTask containerSizeCountTask) {
     super(nodeManager, pipelineManager, containerManager);
     this.scmClient = scmClient;
+    this.containerManager = containerManager;
```
[the remark this answers is not in the document]
Comment (dombizita) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconDeadNodeHandler.java:60]: what was the conclusion here?

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconDeadNodeHandler.java (hunk 11 above)
Comment (dombizita) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconDeadNodeHandler.java:88]: this means that the ContainerSizeCountTask is possible to run more times in each defined interval? if we are calling the `process()` here too.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconIncrementalContainerReportHandler.java (hunk 13 above)
Comment (dombizita) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconIncrementalContainerReportHandler.java:106]: here the same: this means that the ContainerSizeCountTask is possible to run more times in each defined interval? if we are calling the `process()` here too.

File: hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/tasks/TestContainerSizeCountTask.java [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+import org.hadoop.ozone.recon.schema.UtilizationSchemaDefinition;
+import org.hadoop.ozone.recon.schema.tables.daos.ContainerCountBySizeDao;
+import org.hadoop.ozone.recon.schema.tables.daos.ReconTaskStatusDao;
+import org.jooq.DSLContext;
+import org.jooq.Record1;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Class to test a process and reprocess methods of ContainerSizeCountTask.
```
Comment (dombizita) [inline hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/tasks/TestContainerSizeCountTask.java:44]: please update documentation as there is no reprocess method anymore

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java [hunk 20]
```diff
@@ -0,0 +1,367 @@
...
+    this.containerCountBySizeDao = containerCountBySizeDao;
+    this.dslContext = utilizationSchemaDefinition.getDSLContext();
+    interval = reconTaskConfig.getContainerSizeCountTaskInterval().toMillis();
+  }
+
+
+  /**
+   * The run() method is the main loop of the ContainerSizeCountTask class.
+   * It periodically retrieves a list of containers from the containerManager,
+   * and then calls either to reprocess() or process() method depending on
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java:84]: please update documentation as there is no reprocess method anymore

**Review by dombizita — COMMENTED** [review]
thank you so much for working on this @ArafatKhan2198 and for providing so much information about the implementation in the description and in the docs attached, it made the review much easier! I have a few small questions/comments, please take a look at them

File: hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/persistence/TestUtilizationSchemaDefinition.java [hunk 21]
```diff
@@ -103,6 +104,21 @@ public void testReconSchemaCreated() throws Exception {
         "Unexpected number of columns");
     assertEquals(expectedPairsFileCount, actualPairsFileCount,
         "Columns Do not Match ");
+
+
+    ResultSet resultSetContainerCount = metaData.getColumns(null, null,
+        CONTAINER_COUNT_BY_SIZE_TABLE_NAME, null);
+
```
[the remark this answers is not in the document]
Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/persistence/TestUtilizationSchemaDefinition.java:111]: Good find @dombizita Ill make the changes !!
Thanks!

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ReconTaskConfig.java [hunk 22]
```diff
@@ -83,4 +83,23 @@ public Duration getSafeModeWaitThreshold() {
   public void setSafeModeWaitThreshold(Duration safeModeWaitThreshold) {
     this.safeModeWaitThreshold = safeModeWaitThreshold.toMillis();
   }
+
+  @Config(key = "containercounttask.interval",
+      type = ConfigType.TIME,
+      defaultValue = "60s",
+      tags = { ConfigTag.RECON, ConfigTag.OZONE },
+      description = "The time interval of the periodic check for " +
+          "container size distribution in the cluster as reported "
```
[the remark this answers is not in the document]
Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ReconTaskConfig.java:92]: Good find @dombizita Ill make the changes !!
Thanks!

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/UtilizationEndpoint.java [hunk 23]
```diff
@@ -98,4 +111,44 @@ public Response getFileCounts(
...
+  public Response getContainerCounts(
+      @QueryParam(RECON_QUERY_CONTAINER_SIZE)
+          long containerSize) {
+    DSLContext dslContext = utilizationSchemaDefinition.getDSLContext();
+    Long containerSizeUpperBound =
+        ReconUtils.getContainerSizeUpperBound(containerSize);
+    List<ContainerCountBySize> resultSet;
+    try {
+      if (containerSize > 0) {
+        // Get the current count from database and update
```
[the remark this answers is not in the document]
Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/UtilizationEndpoint.java:131]: I intended to update the previous count displayed to the user through the endpoint, but it seems unnecessary to add it to the message. So Ill change it 
Thanks for the find!

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/UtilizationEndpoint.java (hunk 17 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/UtilizationEndpoint.java:141]: I don't think we need to consider containers with a size of zero, as they likely don't exist. Additionally, including the count of zero-sized containers in the graph may not be meaningful.

**Comment by ArafatKhan2198** [comment]
https://github.com/apache/ozone/pull/4107#discussion_r1166750547

So the DeadNodeHandler calls the `removeContainerReplicas` method, which is responsible for removing the container replicas associated with the dead datanode, unless it is in maintenance. Hence, since the container replicas are removed, the container size counts will also be affected. Therefore, we need to trigger the ContainerSizeCountTask, and hence the ContainerSizeCountTask is possible to run more times in each defined interval.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconIncrementalContainerReportHandler.java (hunk 13 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconIncrementalContainerReportHandler.java:106]: Yes @dombizita it will run more number of times then !!

File: hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/tasks/TestContainerSizeCountTask.java (hunk 19 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/tasks/TestContainerSizeCountTask.java:44]: Sure will change!!

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java (hunk 20 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerSizeCountTask.java:84]: Sure will change!!

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconDeadNodeHandler.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconDeadNodeHandler.java:60]: We needed to trigger the process method of the `containerSizeCountTask` from the `ReconDeadNodeHandler`  hence we need to **pass in a updated list of existing containers** to the process method, so as to update the `containerCounts` of those containers that are no more.

**Comment by ArafatKhan2198** [comment]
So after a few discussions with @nandakumar131
The conversation was about adding a trigger to the `IncrementalContainerReportHandler` in order to trigger the `ContainerCountTask` whenever the handler is executed. It was determined that the `ICRReportHandler` will be triggered frequently in a busy cluster, and adding a trigger would put additional load on the cluster. Therefore, the final decision was not to add the trigger to the handler.

@devmadhuu @dombizita please take a final look at this !!

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconIncrementalContainerReportHandler.java [hunk 24]
```diff
@@ -46,7 +46,7 @@
       ReconIncrementalContainerReportHandler.class);
 
   public ReconIncrementalContainerReportHandler(NodeManager nodeManager,
-             ContainerManager containerManager, SCMContext scmContext) {
+                    ContainerManager containerManager, SCMContext scmContext) {
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconIncrementalContainerReportHandler.java:49]: I think we should remove this unnecessary change in the class

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconStorageContainerManagerFacade.java [hunk 25]
```diff
@@ -165,9 +174,7 @@ public ReconStorageContainerManagerFacade(OzoneConfiguration conf,
     this.eventQueue = new EventQueue();
     eventQueue.setSilent(true);
     this.scmContext = new SCMContext.Builder()
-        .setIsPreCheckComplete(true)
-        .setSCM(this)
-        .build();
+        .setIsPreCheckComplete(true).setSCM(this).build();
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconStorageContainerManagerFacade.java:177]: I think we should remove this unnecessary change in the class (also some others in this class as well)

**Comment by dombizita** [comment]
> So after a few discussions with @nandakumar131 The conversation was about adding a trigger to the `IncrementalContainerReportHandler` in order to trigger the `ContainerCountTask` whenever the handler is executed. It was determined that the `ICRReportHandler` will be triggered frequently in a busy cluster, and adding a trigger would put additional load on the cluster. Therefore, the final decision was not to add the trigger to the handler.
> 
> @devmadhuu @dombizita please take a final look at this !!

during the discussion did you talk about [ReconDeadNodeHandler.java](https://github.com/apache/ozone/pull/4107/files#diff-cf88738ad7edb0171749525ea362b6172a30c36e206e487d0ed916e226c8927d)? do you think it is good to trigger the task beside the periodic runs, is it helping that much in the understanding of container sizes? won't it cause frequent calls in some cases? 

beside this I am fine with your changes, I added a few small comments to remove unnecessary code changes in some classes. I'd like to have @devmadhuu to review it as well as I am not sure if all of his previous comments have been addressed.

**Comment by ArafatKhan2198** [comment]
> > So after a few discussions with @nandakumar131 The conversation was about adding a trigger to the `IncrementalContainerReportHandler` in order to trigger the `ContainerCountTask` whenever the handler is executed. It was determined that the `ICRReportHandler` will be triggered frequently in a busy cluster, and adding a trigger would put additional load on the cluster. Therefore, the final decision was not to add the trigger to the handler.
> > @devmadhuu @dombizita please take a final look at this !!
> 
> during the discussion did you talk about [ReconDeadNodeHandler.java](https://github.com/apache/ozone/pull/4107/files#diff-cf88738ad7edb0171749525ea362b6172a30c36e206e487d0ed916e226c8927d)? do you think it is good to trigger the task beside the periodic runs, is it helping that much in the understanding of container sizes? won't it cause frequent calls in some cases?
> 
> beside this I am fine with your changes, thanks for addressing some of them and answering my questions @ArafatKhan2198! I added a few small comments to remove unnecessary code changes in some classes. I'd like to have @devmadhuu to review it as well as I am not sure if all of his previous comments have been addressed.

Thank you for your comment, @dombizita . I appreciate your feedback and concerns regarding the ReconDeadNodeHandler.java and its impact on container size understanding. However, I believe that the frequency of a node going down in a busy cluster is not very high, so triggering the task beside the periodic runs should not significantly impact performance. Additionally, I would like to inform you that @devmadhuu  has already discussed with me the pending  issues in an offline discussion, and he is good to go on this!

I will review your code suggestions and make any necessary changes. Thank you for taking the time to review my code.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconIncrementalContainerReportHandler.java (hunk 24 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconIncrementalContainerReportHandler.java]: Removed thanks !!

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconStorageContainerManagerFacade.java (hunk 25 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconStorageContainerManagerFacade.java]: Yea I have removed this and and the others as well!!

**Review by dombizita — APPROVED** [review]
thank you @ArafatKhan2198 for your great work on this patch! it looks good to me!

**Review by devmadhuu — APPROVED** [review]

**Comment by smengcl** [comment]
Thanks @ArafatKhan2198 for the improvement, @devmadhuu @dombizita for reviewing this.
